cli/2024-06-04-aws-irsa-support.md

```diff
@@ -0,0 +1,326 @@
+ 
+```
+rad credential register aws --access-key-id <access-key-id> --secret-access-key <secret-access-key>
+```
+ 
+These IAM credentials, including the access key should be rotated regularly to reduce the chance of unauthorized access. A more secure option than using these credentials is to use IRSA.
+ 
+IRSA (IAM Roles for Service Accounts) is used for authenticating applications running within Kubernetes pods. When we use IRSA, we associate an IAM role directly with a Kubernetes service-account, allowing the pods associated with the service-account to assume that role. IRSA does not require us to configure access key and  secret. It relies identity tokens (OIDC tokens) to authenticate with AWS services. These tokens have a default expiration (configurable). However, they automatically refresh when needed, transparent to the user.
+
+The goal of the scenario is to enable infrastructure operators to configure IRSA support for the AWS provider in Radius to deploy and manage AWS resources.
+
+## Terms and definitions
+
+
+## Objectives
+
+> Issue Reference: https://github.com/radius-project/radius/issues/7618 
+
+### Goals
+
+* Radius users can configure AWS provider and enable Radius to use IRSA for authentication and deployment of application.
+* IRSA can be configured via interactive experience.
+* IRSA can be configured manually.
+* Radius users can deploy and manage AWS resources without having to manage AWS access key id and secret
+
+### Non-goals
+
+* Azure Managed Identity support
+* Azure Workload Identity support
+* Ability of radified applications to be able to use IRSA for authenticating AWS operations.
+
+### User scenarios
+
+As a user, I should be able to configure Radius with IRSA for managing AWS resources.
+
+## Design
+
+### High Level Design for IRSA
+
+During installation, Radius should allow user to enable IRSA for AWS while conifguring AWS provider. Then the user should be able to enter the roleARN associated with desired policies which will be stored as the AWS credential.
+
+At a high level, the full flow to setup IRSA looks as below:
+
+1. configure the cluster with oidc provider. have the role-arn configured with desired policies ready.
+2. take steps to allow  `rad-role-arn` and `eks cluster : namespace: service` to trust each other using aws portal.
+3. use  `rad init` to configure radius with role-arn credential. (or use rad credential later on)
+4. use `rad env update` to configure provider associated with the specific environment (`rad env update qa --aws-account-id <aws-account-id> --aws-region <aws-region>`). 
+5. during rad deploy, radius assumes the role that is configured in the credential. 
+
+Since oidc is configured for a cluster and not specific to a role-arn, the solution can be extended to work in multi-tenancy once we have multi credentials support. 
+
+More details about each step is covered in [Detailed Design](#detailed-design)
+
+IRSA works the same way on EKS as well as non-EKS clusters.  
```

Comment:
What is the expected IAM policy for non-EKS clusters?

Reply:
the IAM policy should allow Radius to manage resources we support and follow the principle of least privilege as good practice. It would be the same on both eks and non-eks cluster.

Reply:
I was trying to get more specific details about what the policy would look like similar to the example you added here https://github.com/radius-project/design-notes/blob/8aca2d8a3b550ed8f1035d0ce57c2034999ed4a0/cli/2024-06-04-aws-irsa-support.md?plain=1#L164, to understand it better. :) Feel free to share a link to AWS docs if it is already documented on EKS docs.

Reply:
The Principal looks generally similar to  
```
"Principal": {
          "Federated": "arn:aws:iam::${ACCOUNT_ID}:oidc-provider/${OIDC_PROVIDER}"
        },
```
This is applicable if we use a AWS OIDC provider, irrespective of the cluster.  If we want another OIDC provider, https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html covers steps for configuring that, and them using the IAM of that provider.